api/task_manager.cc

```diff
@@ -0,0 +1,163 @@
+    ::tm st, et;
+    ::gmtime_r(&end_time, &et);
+    ::gmtime_r(&start_time, &st);
+
+    tm::task_status res{};
+    res.id = status.task_status.id.to_sstring();
+    res.type = status.task_status.type;
+    res.state = status.task_status.state;
+    res.is_abortable = bool(status.abortable);
+    res.start_time = st;
+    res.end_time = et;
+    res.error = status.task_status.error;
+    res.parent_id = status.parent_id.to_sstring();
+    res.sequence_number = status.task_status.sequence_number;
+    res.shard = status.task_status.shard;
+    res.keyspace = status.task_status.keyspace;
+    res.table = status.task_status.table;
+    res.entity = status.task_status.entity;
+    res.progress_units = status.task_status.progress_units;
+    res.progress_total = status.progress.total;
+    res.progress_completed = status.progress.completed;
+    return res;
+}
+
+future<json::json_return_type> retrieve_status(task_manager::foreign_task_ptr task) {
+    if (task.get() == nullptr) {
+        return make_exception_future<json::json_return_type>(httpd::bad_param_exception("Invalid task id"));
+    }
+    full_task_status s;
+    s.task_status = task->get_status();
+    s.parent_id = task->get_parent_id();
+    s.abortable = task->is_abortable();
+    s.module = task->get_module()->get_name();
+    return task->get_progress().then([s] (task_manager::task::progress progress) mutable {
+        s.progress.completed = progress.completed;
+        s.progress.total = progress.total;
+        return make_ready_future<json::json_return_type>(make_status(s));
+    });
+}
+
+void set_task_manager(http_context& ctx, routes& r) {
+    tm::get_modules.set(r, [&ctx] (std::unique_ptr<request> req) {
+        std::vector<std::string> v;
+        for (auto el : ctx.tm.local().get_modules()) {
+            v.push_back(el.first);
+        }
+        return make_ready_future<json::json_return_type>(v);
+    });
+
+    tm::get_tasks.set(r, [&ctx] (std::unique_ptr<request> req) {
+        return do_with(req.get(), [&ctx] (auto &req) {
+            return ctx.tm.map_reduce0([&req] (task_manager& tm) {
+                std::vector<tm::task_stats> res{};
+                auto module = tm.get_modules()[req->param["module"]];
+                if (module) {
+                    const auto& filtered_tasks = module->get_tasks() | boost::adaptors::filtered([&params = req->query_parameters] (const auto& task) {
+                        return filter_tasks(task.second, params);
+                    });
+                    for (auto& [task_id, task] : filtered_tasks) {
+                        tm::task_stats ts;
+                        ts.task_id = task->get_status().id.to_sstring();
+                        ts.state = task->get_status().state;
+                        res.push_back(ts);
+                    }
+                }
+                return res;
+            }, std::vector<tm::task_stats>(), concat<tm::task_stats>);
```

Comment:
I wonder if we can stream the results instead of copying them to an intermediary vector.
It doesn't matter for a small number of tasks but if we get to the 1000's or 10's of thousands it will definitely matter
and you'd want to reduce buffering
@amnonh ?

Reply:
Do you mean that it should be passed to a stringstream with all json-specific symbols (like "[" or ",") being inserted openly at api side?
And, if so, should the return type in `api/task_manager.cc` be changed from an array of task_stats to string?

Reply:
One example (which I'm not sure is the best... I vaguely remember we had a better way to stream an array of objects, but I don't recall exactly where) is in https://github.com/scylladb/scylladb/blob/516089beb0b8960ee5a9b9b73e8664e37e2275ac/api/storage_service.cc#L1277
where the api implementation formats the result into an `output_stream` as an array, and explicitly printing `[`, `]`, and `, ` delimiters.

Reply:
There are more examples that use `stream_range_as_array`, but I think they all work on already-materialized containers like vectors or list.

In this case, to avoid generating such an intermediate vector I'm interested in using a coroutine generator that will yield the results asynchronously, but I'm not sure we're ready for that even on the seastar side of things (Cc @avikivity).